File: Challenges/450_DSA_Cracker/Array/009_Minimise_The_Difference.py
#User function Template for python3

class Solution:
    def getMinDiff(self, arr, n, k):
        arr = sorted(arr)
        n = len(arr)
        mi = arr[0]
        ma = arr[n-1]
        ans = ma - mi
        for i in range(1, n):
            if arr[i] - k < 0:
                continue
            mi = min(arr[0] + k, arr[i] - k)
            ma = max(arr[i-1] + k, arr[n-1] - k)
            ans = min(ans, ma - mi)
        return ans


#{ 
 # Driver Code Starts
#Initial Template for Python 3

if __name__ == '__main__':
    tc = int(input())
    while tc > 0:
        k = int(input())
        n = int(input())
        arr = list(map(int, input().strip().split()))
        ob = Solution()
        ans = ob.getMinDiff(arr, n, k)
        print(ans)
        tc -= 1

# } Driver Code Ends

File: Challenges/450_DSA_Cracker/Array/009_Minimise_The_Difference.java
//{ Driver Code Starts
// Initial Template for Java

import java.io.*;
import java.util.*;

  public class Main {

    public static void main(String[] args) throws Exception {
        BufferedReader br =
            new BufferedReader(new InputStreamReader(System.in));
        int tc = Integer.parseInt(br.readLine().trim());
        while (tc-- > 0) {
            String[] inputLine;
            inputLine = br.readLine().trim().split(" ");
            int k = Integer.parseInt(inputLine[0]);
            inputLine = br.readLine().trim().split(" ");
            int n = Integer.parseInt(inputLine[0]);
            int[] arr = new int[n];
            inputLine = br.readLine().trim().split(" ");
            for (int i = 0; i < n; i++) {
                arr[i] = Integer.parseInt(inputLine[i]);
            }

            int ans = new Solution().getMinDiff(arr, n, k);
            System.out.println(ans);
        }
    }
}
// } Driver Code Ends


// User function Template for Java

class Solution {
    int getMinDiff(int[] arr, int n, int k) {
        Arrays.sort(arr);
        int minHeight = arr[0], maxHeight = arr[n-1];
        int diff = maxHeight - minHeight;
        for(int i=1; i<n; i++) {
            if(arr[i] - k < 0) continue;
            minHeight = Math.min(arr[0] + k, arr[i] - k);
            maxHeight = Math.max(arr[i-1] + k, arr[n-1] - k);
            diff = Math.min(diff, maxHeight - minHeight);
        }
        return diff;
    }
}
